lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,683 @@
+//===- GrandCentralTaps.cpp -------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the GrandCentralTaps pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#define DEBUG_TYPE "gct"
+
+using llvm::SmallDenseMap;
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Utilities
+//===----------------------------------------------------------------------===//
+
+namespace llvm {
+
+// Make InstanceOp behave like the underlying `Operation *`.
+template <>
+struct PointerLikeTypeTraits<InstanceOp> {
+  static inline void *getAsVoidPointer(InstanceOp value) {
+    return const_cast<void *>(value.getAsOpaquePointer());
+  }
+  static inline InstanceOp getFromVoidPointer(void *pointer) {
+    return InstanceOp::getFromOpaquePointer(pointer);
+  }
+  enum {
+    NumLowBitsAvailable =
+        PointerLikeTypeTraits<Operation *>::NumLowBitsAvailable
+  };
+};
+
+// Make DictionaryAttr behave like the underlying `Attribute`.
+template <>
+struct PointerLikeTypeTraits<DictionaryAttr> {
+  static inline void *getAsVoidPointer(DictionaryAttr value) {
+    return const_cast<void *>(value.getAsOpaquePointer());
+  }
+  static inline DictionaryAttr getFromVoidPointer(void *pointer) {
+    return DictionaryAttr::getFromOpaquePointer(pointer);
+  }
+  enum {
+    NumLowBitsAvailable = PointerLikeTypeTraits<Attribute>::NumLowBitsAvailable
+  };
+};
+
+} // namespace llvm
+
+//===----------------------------------------------------------------------===//
+// Static information gathered once upfront
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Attributes used throughout the annotations.
+struct Strings {
+  MLIRContext *const cx;
+  Strings(MLIRContext *cx) : cx(cx) {}
+
+  Identifier annos = Identifier::get("annotations", cx);
+  Identifier fannos = Identifier::get("firrtl.annotations", cx);
+
+  StringAttr dataTapsClass =
+      StringAttr::get(cx, "sifive.enterprise.grandcentral.DataTapsAnnotation");
```

Comment:
Yes. The members are initialized in the order they were declared in the class, and regardless of how the constructor lists them. So the `context(context)` runs first because the `MLIRContext *context` decl appears first in the class.